htap_test.go

```diff
@@ -101,10 +101,161 @@ func trimWhitespaces(s string) string {
 	)
 }
 
+func TestSortingByPriority(t *testing.T) {
+	qcc := (&queryContextCache{}).init()
+
+	qceA := queryContextEntry{ID: 12, Timestamp: 123, Priority: 7, Context: "a"}
+	qceB := queryContextEntry{ID: 13, Timestamp: 124, Priority: 9, Context: "b"}
+	qceC := queryContextEntry{ID: 14, Timestamp: 125, Priority: 6, Context: "c"}
+	qceD := queryContextEntry{ID: 15, Timestamp: 126, Priority: 8, Context: "d"}
+
+	t.Run("Add to empty cache", func(t *testing.T) {
+		qcc.add(qceA)
+		if !reflect.DeepEqual(qcc.entries, []queryContextEntry{qceA}) {
+			t.Fatalf("no entries added to cache. %v", qcc.entries)
+		}
+	})
+	t.Run("Add another entry with different id, timestamp and priority - greater priority", func(t *testing.T) {
+		qcc.add(qceB)
+		if !reflect.DeepEqual(qcc.entries, []queryContextEntry{qceA, qceB}) {
+			t.Fatalf("entry with greater priority should be added at the end. %v", qcc.entries)
+		}
+	})
+	t.Run("Add another entry with different id, timestamp and priority - lesser priority", func(t *testing.T) {
+		qcc.add(qceC)
+		if !reflect.DeepEqual(qcc.entries, []queryContextEntry{qceC, qceA, qceB}) {
+			t.Fatalf("entry with lesser priority should be added at the beginninig. %v", qcc.entries)
+		}
+	})
+	t.Run("Add another entry with different id, timestamp and priority - priority in the middle", func(t *testing.T) {
+		qcc.add(qceD)
+		if !reflect.DeepEqual(qcc.entries, []queryContextEntry{qceC, qceA, qceD, qceB}) {
+			t.Fatalf("entry with priority in the middle should be added at the middle. %v", qcc.entries)
+		}
+	})
+}
+
+func TestAddingQcesWithTheSameIdAndLaterTimestamp(t *testing.T) {
+	qcc := (&queryContextCache{}).init()
+
+	qceA := queryContextEntry{ID: 12, Timestamp: 123, Priority: 7, Context: "a"}
+	qceB := queryContextEntry{ID: 13, Timestamp: 124, Priority: 9, Context: "b"}
+	qceC := queryContextEntry{ID: 12, Timestamp: 125, Priority: 6, Context: "c"}
+	qceD := queryContextEntry{ID: 12, Timestamp: 126, Priority: 6, Context: "d"}
+
+	t.Run("Add to empty cache", func(t *testing.T) {
+		qcc.add(qceA)
+		qcc.add(qceB)
+		if !reflect.DeepEqual(qcc.entries, []queryContextEntry{qceA, qceB}) {
+			t.Fatalf("no entries added to cache. %v", qcc.entries)
+		}
+	})
+	t.Run("Add another entry with different priority", func(t *testing.T) {
+		qcc.add(qceC)
+		if !reflect.DeepEqual(qcc.entries, []queryContextEntry{qceC, qceB}) {
+			t.Fatalf("entry with greater priority should be added at the end. %v", qcc.entries)
```

Comment:
so why it is added at the beggining? and maybe it is worth adding "and should replace entry with same ID"?